dds/DCPS/XTypes/TypeLookupService.cpp

```diff
@@ -454,6 +473,425 @@ bool TypeLookupService::complete_to_minimal_type_object(const TypeObject& cto, T
+    td.bound[0] = ti.string_sdefn().bound;
+    td.element_type = type_identifier_to_dynamic(TypeIdentifier(TK_CHAR8), dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_STRING16_SMALL:
+    td.kind = TK_STRING16;
+    td.name = "WString16Small";
+    td.bound.length(1);
+    td.bound[0] = ti.string_ldefn().bound;
+    td.element_type = type_identifier_to_dynamic(TypeIdentifier(TK_CHAR16), dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_STRING16_LARGE:
+    td.kind = TK_STRING16;
+    td.name = "WString16Large";
+    td.bound.length(1);
+    td.bound[0] = ti.string_ldefn().bound;
+    td.element_type = type_identifier_to_dynamic(TypeIdentifier(TK_CHAR16), dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_PLAIN_SEQUENCE_SMALL:
+    td.kind = TK_SEQUENCE;
+    td.name = "SequenceSmall";
+    td.bound.length(1);
+    td.bound[0] = ti.seq_sdefn().bound;
+    td.element_type = type_identifier_to_dynamic(*ti.seq_sdefn().element_identifier, dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_PLAIN_SEQUENCE_LARGE:
+    td.kind = TK_SEQUENCE;
+    td.name = "SequenceLarge";
+    td.bound.length(1);
+    td.bound[0] = ti.seq_ldefn().bound;
+    td.element_type = type_identifier_to_dynamic(*ti.seq_ldefn().element_identifier, dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_PLAIN_ARRAY_SMALL:
+    td.kind = TK_ARRAY;
+    td.name = "ArraySmall";
+    td.bound.length(ti.array_sdefn().array_bound_seq.length());
+    for (ACE_CDR::ULong i = 0; i< td.bound.length(); ++i) {
+      td.bound[i] = ti.array_sdefn().array_bound_seq[i];
+    }
+    td.element_type = type_identifier_to_dynamic(*ti.array_sdefn().element_identifier, dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_PLAIN_ARRAY_LARGE:
+    td.kind = TK_ARRAY;
+    td.name = "ArrayLarge";
+    td.bound = ti.array_ldefn().array_bound_seq;
+    td.element_type = type_identifier_to_dynamic(*ti.array_ldefn().element_identifier, dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_PLAIN_MAP_SMALL:
+    td.kind = TK_MAP;
+    td.name = "MapSmall";
+    td.bound.length(1);
+    td.bound[0] = ti.map_sdefn().bound;
+    td.element_type = type_identifier_to_dynamic(*ti.map_sdefn().element_identifier, dt_map);
+    td.key_element_type = type_identifier_to_dynamic(*ti.map_sdefn().element_identifier, dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_PLAIN_MAP_LARGE:
+    td.kind = TK_MAP;
+    td.name = "MapLarge";
+    td.bound.length(1);
+    td.bound[0] = ti.map_ldefn().bound;
+    td.element_type = type_identifier_to_dynamic(*ti.map_ldefn().element_identifier, dt_map);
+    td.key_element_type = type_identifier_to_dynamic(*ti.map_ldefn().element_identifier, dt_map);
+    dt->set_descriptor(td);
+    break;
+  case TI_STRONGLY_CONNECTED_COMPONENT:
+    complete_to_dynamic_i(dt, get_type_object_i(ti).complete, dt_map);
+    break;
+  case EK_COMPLETE:
+    complete_to_dynamic_i(dt, get_type_object_i(ti).complete, dt_map);
+    break;
```

Comment:
I break that infinite loop by using the `DynamicTypeMap` which checks to see if I have already encountered the DynamicType. Recent changes made it so I had to move the creation of `DynamicTypeMap` to outside where I call into `complete_to_dynamic` so that the weak_rch don't destroy themselves early. Any time you are using `complete_to_dynamic` you will have to do the same.

In the spec issue document I think I already created an entry for the infinite looping problem.

Reply:
Edit: I reread what I put for the spec issue and that writeup I did about infinite recursion is about checking equality, not about what is discussed here.

We could discuss with team tomorrow about whether this is more of an implementation choice vs whether it is a spec issue.